compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java

```diff
@@ -3003,76 +3011,315 @@ private BLangExpression createConditionFromMatchPatterns(List<BLangMatchPattern>
-                ASTBuilderUtil.createAssignmentStmt(pos, resultVarRef, getBooleanLiteral(false));
-        failureBody.addStatement(failureAssignment);
+        BLangBlockStmt successBody = createSuccessOrFailureBody(true, resultVarRef, pos);
 
         BLangIf parentIfElse = createIfElseStmtFromMatchPattern(matchPatterns.get(0), matchExprVar, successBody, pos);
         BLangIf currentIfElse = parentIfElse;
 
         for (int i = 1; i < matchPatterns.size(); i++) {
+            successBody = createSuccessOrFailureBody(true, resultVarRef, pos);
             currentIfElse.elseStmt = createIfElseStmtFromMatchPattern(matchPatterns.get(i), matchExprVar, successBody,
                     matchPatterns.get(i).pos);
             currentIfElse = (BLangIf) currentIfElse.elseStmt;
         }
-        currentIfElse.elseStmt = failureBody;
 
+        currentIfElse.elseStmt = createSuccessOrFailureBody(false, resultVarRef, pos);
         BLangBlockStmt blockStmt = ASTBuilderUtil.createBlockStmt(pos, Lists.of(resultVarDef, parentIfElse));
         BLangStatementExpression stmtExpr = createStatementExpression(blockStmt, resultVarRef);
 
         return rewriteExpr(stmtExpr);
     }
 
+    private BLangBlockStmt createSuccessOrFailureBody(boolean status, BLangSimpleVarRef varRef, DiagnosticPos pos) {
+        BLangBlockStmt blockStmt = ASTBuilderUtil.createBlockStmt(pos);
+        BLangAssignment bLangAssignment =
+                ASTBuilderUtil.createAssignmentStmt(pos, varRef, getBooleanLiteral(status));
+        blockStmt.addStatement(bLangAssignment);
+        return blockStmt;
+    }
+
     private BLangIf createIfElseStmtFromMatchPattern(BLangMatchPattern matchPattern,
-                                                             BLangSimpleVariable matchExprVar,
-                                                             BLangBlockStmt successBody,
-                                                             DiagnosticPos pos) {
-        NodeKind patternKind = matchPattern.getKind();
+                                                     BLangSimpleVariable matchExprVar,
+                                                     BLangBlockStmt successBody,
+                                                     DiagnosticPos pos) {
+
         BLangSimpleVarRef matchExprVarRef = ASTBuilderUtil.createVariableRef(matchExprVar.pos, matchExprVar.symbol);
 
+        BLangExpression condition = createConditionForMatchPattern(matchPattern, matchExprVarRef);
+        successBody.getStatements().addAll(matchStmtsForPattern);
+        matchStmtsForPattern.clear();
+        return ASTBuilderUtil.createIfElseStmt(pos, condition, successBody, null);
+    }
+
+    private BLangExpression createConditionForWildCardMatchPattern(BLangWildCardMatchPattern wildCardMatchPattern,
+                                                                   BLangSimpleVarRef matchExprVarRef) {
+
+        boolean condition;
+        if (wildCardMatchPattern.matchesAll) {
```

Comment:
will it be possible to collapse the if statement to a conditional op?

Reply:
Done